src/purerpc/grpc_socket.py

```diff
@@ -231,8 +231,9 @@ def _allocate_stream(self, stream_id):
 
     async def _listen(self):
         while True:
-            data = await self._socket.recv(self._receive_buffer_size)
-            if not data:
+            try:
+                data = await self._socket.recv(self._receive_buffer_size)
+            except (anyio.EndOfStream, anyio.BrokenResourceError):
```

Comment:
I manually confirmed that BrokenResourceError is also leaking to the server main loop since the anyio upgrade.  Before the upgrade, no tests triggered exceptions in the main loop, and after both EndOfStream and BrokenResourceError are.

Does the presence of BrokenResourceError indicate another problem in the code?  I'm not too knowledgeable about these details-- what makes you concerned?

Reply:
Maybe BrokenResourceError should be handled, but at some other level?

Reply:
I'm not knowledgeable about the protocol itself, but `EndOfStream` is raised when the remote has closed the stream and there is nothing more to read. `BrokenResourceError`, however, is raised, for example, if a send is attempted on a closed socket. This can occur even during an operation that is superficially a receive, like with a TLS connection. This error usually indicates that the peer closed the connection prematurely. If this is the kind of error you want to ignore here, then that's fine.

Reply:
If it's not caught here, there are only a few other opportunities.

The cause of the BrokenResourceError is `ConnectionResetError: [Errno 54] Connection reset by peer`.

  File "/Users/john/.pyenv/versions/3.8.12/lib/python3.8/asyncio/selector_events.py", line 848, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
ConnectionResetError: [Errno 54] Connection reset by peer

The above exception was the direct cause of the following exception:

```
  File "/.../asyncio/selector_events.py", line 848, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
ConnectionResetError: [Errno 54] Connection reset by peer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/.../purerpc/src/purerpc/server.py", line 221, in __call__
    async for stream in self.grpc_socket.listen():
  File "/.../purerpc/src/purerpc/grpc_socket.py", line 265, in listen
    async for value in self._listen():
  File "/.../purerpc/src/purerpc/grpc_socket.py", line 234, in _listen
    data = await self._socket.recv(self._receive_buffer_size)
  File "/.../purerpc/src/purerpc/grpc_socket.py", line 72, in recv
    return await self._stream.receive(buffer_size)
  File "/.../purerpc/venv/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 1114, in receive
    raise self._protocol.exception
anyio.BrokenResourceError
```

Reply:
added a note in the code